### PR TITLE
[mlir][ArmSVE] Lower predicate-sized vector.create_masks to whilelt

### PR DESCRIPTION
**Pull request by MacDue**
This produces better/more canonical codegen than the generic LLVM lowering, which is a pattern the backend currently does not recognize. See: https://github.com/llvm/llvm-project/issues/81840.

### CODE REVIEW

**Comment by MacDue** [comment]
As an example:

```
func.func @mask(%a: index) -> vector<[16]xi1> {
  %0 = vector.create_mask %a : vector<[16]xi1>
  return %0 : vector<[16]xi1>
}
```

Currently lowers to:
```
	index	z0.s, #0, #1
	mov	z1.s, w0
	ptrue	p0.s
	mov	z2.d, z0.d
	mov	z3.d, z0.d
	cmpgt	p3.s, p0/z, z1.s, z0.s
	incw	z2.s
	incw	z3.s, all, mul #2
	mov	z4.d, z2.d
	cmpgt	p1.s, p0/z, z1.s, z3.s
	incw	z4.s, all, mul #2
	cmpgt	p2.s, p0/z, z1.s, z4.s
	cmpgt	p0.s, p0/z, z1.s, z2.s
	uzp1	p1.h, p1.h, p2.h
	uzp1	p0.h, p3.h, p0.h
	uzp1	p0.b, p0.b, p1.b
	ret
```

With this patch it lowers to:
```
	whilelt	p0.b, xzr, x0
	ret
```

**Review by c-rhodes — COMMENTED** [review]
thanks for the patch Ben this is a nice improvement!

Left some comments mostly minor, also occurred to me this is the first Vector -> ArmSVE conversion but we're doing it during LLVM conversion. I don't think we want to prematurely introduce a vector-to-arm-sve conversion pass, but something to bear in mind if this grows.

**Comment by MacDue** [comment]
> Left some comments mostly minor, also occurred to me this is the first Vector -> ArmSVE conversion but we're doing it during LLVM conversion. 

It's Vector -> LLVM SVE intrinsics (not high-level ArmSVE operations). I'm not sure if a general Vector -> ArmSVE conversion would make much sense, as I don't think it'd be much more expressive than the vector dialect (and there's no problems needing solving there, e.g. like tile allocation for SME). Also, with some later patches in mind the ArmSVE ops I use have specific requirements (like `+sme` or `+sve2`), so a general conversion pass would be fairly limited in what it could do.

**Review by c-rhodes — APPROVED** [review]

File: mlir/lib/Dialect/ArmSVE/Transforms/LegalizeForLLVMExport.cpp [hunk 1]
```diff
@@ -140,6 +140,40 @@ using ConvertFromSvboolOpLowering =
 using ZipX2OpLowering = OneToOneConvertToLLVMPattern<ZipX2Op, ZipX2IntrOp>;
 using ZipX4OpLowering = OneToOneConvertToLLVMPattern<ZipX4Op, ZipX4IntrOp>;
 
+/// Converts `vector.create_mask` ops that match the size of an SVE predicate
+/// to the `whilelt` intrinsic. This produces more canonical codegen than the
+/// generic LLVM lowering, see https://github.com/llvm/llvm-project/issues/81840
+/// for more details. Note that we can't (the more general) get.active.lane.mask
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSVE/Transforms/LegalizeForLLVMExport.cpp:146]:
```suggestion
/// for more details. Note that we can't use (the more general) get.active.lane.mask
```
?

**Review by banach-space — APPROVED** [review]
Nice, thanks!

What was the lowering path before? As in, we'd start with ` vector.create_mask` and then ...?

**Comment by MacDue** [comment]
> What was the lowering path before? As in, we'd start with  vector.create_mask and then ...?

It's shown in the linked issue (`cmp(stepvector, splat)`)

**Pull request merged**